File: adventofcode/2022/day14/day14.rb
# frozen_string_literal: true

def read_input
  ARGF.readlines
end

def parse_input lines
  lines.map {|l| l.split(" -> ").map {|p| p.split(",").map(&:to_i)}}
end

def full_path line
  line.each_cons(2).reduce({}) do |s, (p1, p2)|
    x1, y1 = p1
    x2, y2 = p2

    if x1 == x2 # vertical line
      if y1 < y2
        (y1..y2).each {|yi| s[[x1, yi]] = '#'}
      else
        (y2..y1).each {|yi| s[[x1, yi]] = '#'}
      end
    else        # horizontal line
      if x1 < x2
        (x1..x2).each {|xi| s[[xi, y1]] = '#'}
      else
        (x2..x1).each {|xi| s[[xi, y1]] = '#'}
      end
    end
    s
  end
end

def all_paths path_lines
  path_lines.reduce({}) do |s, ph|
    s.merge(ph)
  end
end

def falling_forever? paths, point
  ymax = paths.keys.map(&:last).max
  _, y = point
  return true if y > ymax
  false
end

def down p
  x, y = p
  [x, y + 1]
end

def left p
  x, y = p
  [x - 1, y]
end

def right p
  x, y = p
  [x + 1, y]
end

def visualize_paths paths, update
  xmin, xmax = paths.keys.map(&:first).minmax
  ymin, ymax = paths.keys.map(&:last).minmax
  ymin = [0, ymin].min
  s = ""
  (ymin..ymax).each do |y|
    (xmin..xmax).each do |x|
      if [x, y] == [500, 0]
        s += '+'
      else
        s += paths[[x, y]] || '.'
      end
    end
    s += "\n"
  end
  puts s
  if update
    sleep (0.01)
    system 'clear'
  end
end

def is_not_wall? paths, point
  paths[point] != '#' && paths[point] != 'o'
end

def simulate_sand paths
  sp = [500, 0]
  highest_resting_sand = [500, 0]
  while true
    return if falling_forever?(paths, sp)

    np = down sp
    sp = if is_not_wall? paths, np
           np
         else
           if is_not_wall? paths, left(np)
             left np
           elsif is_not_wall? paths, right(np)
             right np
           else
             paths[sp] = 'o' # mark as wall
             if highest_resting_sand[1] > sp[1]
               highest_resting_sand[1] = sp[1]
             end
             [500, highest_resting_sand[1] - 1]
           end
         end
    # visualize_paths paths, true
  end
end

def part1 paths
  simulate_sand paths
  paths.count {|_, v| v == 'o'}
end

def visualize_paths_with_floor paths, update
  xmin, xmax = paths.keys.map(&:first).minmax
  ymin, ymax = paths.keys.map(&:last).minmax
  floory = ymax + 2
  ymin = [0, ymin].min
  s = ""
  (ymin..ymax).each do |y|
    (xmin..xmax).each do |x|
      if [x, y] == [500, 0]
        s += '+'
      elsif y == floory
        s += '#'
      else
        s += paths[[x, y]] || '.'
      end
    end
    s += "\n"
  end
  puts s
  if update
    sleep (0.01)
    system 'clear'
  end
end

def is_not_wall_or_floor? paths, floory, point
  return false if point[1] == floory
  paths[point] != '#' && paths[point] != 'o'
end

def simulate_sand_with_floor paths
  sp = [500, 0]
  highest_resting_sand = [500, 0]
  floory = paths.keys.map(&:last).max + 2
  while true
    # return cnt if falling_forever?(paths, sp)
    return if paths[[500, 0]] == 'o'

    np = down sp
    sp = if is_not_wall_or_floor? paths, floory, np
           np
         else
           if is_not_wall_or_floor? paths, floory, left(np)
             left np
           elsif is_not_wall_or_floor? paths, floory, right(np)
             right np
           else
             paths[sp] = 'o' # mark as wall
             if highest_resting_sand[1] > sp[1]
               highest_resting_sand[1] = sp[1]
             end
             [500, highest_resting_sand[1] - 1]
           end
         end
  end
end


def part2 paths
  simulate_sand_with_floor paths
  paths.count {|_, v| v == 'o'}
end

path_points = parse_input read_input
paths = all_paths path_points.map(&method(:full_path))
p part1 paths
p part2 paths

File: adventofcode/2022/day14/day14.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func readInput() []string {
	lines := []string{}
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		lines = append(lines, scanner.Text())
	}
	return lines
}

type Point struct {
	x, y int
}

func (p Point) down() Point {
	return Point{x: p.x, y: p.y + 1}
}
func (p Point) left() Point {
	return Point{x: p.x - 1, y: p.y}
}
func (p Point) right() Point {
	return Point{x: p.x + 1, y: p.y}
}

func parsePaths(lines []string) map[Point]rune {
	paths := make(map[Point]rune)
	for _, line := range lines {
		coords := strings.Split(line, " -> ")
		points := []Point{}
		for _, coord := range coords {
			xys := strings.Split(coord, ",")
			x, _ := strconv.Atoi(xys[0])
			y, _ := strconv.Atoi(xys[1])
			p := Point{x, y}
			points = append(points, p)
		}

		for i := 0; i < len(points)-1; i++ {
			p1 := points[i]
			p2 := points[i+1]
			if p1.x == p2.x {
				if p1.y < p2.y {
					for yi := p1.y; yi <= p2.y; yi++ {
						p := Point{x: p1.x, y: yi}
						paths[p] = '#'
					}
				} else {
					for yi := p2.y; yi <= p1.y; yi++ {
						p := Point{x: p1.x, y: yi}
						paths[p] = '#'
					}
				}
			} else {
				if p1.x < p2.x {
					for xi := p1.x; xi <= p2.x; xi++ {
						p := Point{x: xi, y: p1.y}
						paths[p] = '#'
					}
				} else {
					for xi := p2.x; xi <= p1.x; xi++ {
						p := Point{x: xi, y: p1.y}
						paths[p] = '#'
					}
				}
			}
		}
	}
	return paths
}

func fallingForever(paths map[Point]rune, p Point) bool {
	ymax := 0
	for pi := range paths {
		if ymax < pi.y {
			ymax = pi.y
		}
	}
	return p.y > ymax
}

func isNotWall(paths map[Point]rune, p Point) bool {
	_, ok := paths[p]
	return !ok
}

func simulateSand(paths map[Point]rune) {
	ymin := 0
	p := Point{x: 500, y: 0}
	for {
		if fallingForever(paths, p) {
			return
		}

		np := p.down()
		if isNotWall(paths, np) {
			p = np
		} else {
			if isNotWall(paths, np.left()) {
				p = np.left()
			} else if isNotWall(paths, np.right()) {
				p = np.right()
			} else {
				paths[p] = 'o'
				if ymin > p.y {
					ymin = p.y
				}
				p = Point{x: 500, y: ymin - 1}
			}
		}
	}
}

func countResting(paths map[Point]rune) int {
	cnt := 0
	for _, c := range paths {
		if c == 'o' {
			cnt += 1
		}
	}
	return cnt
}

func part1(paths map[Point]rune) int {
	simulateSand(paths)
	return countResting(paths)
}

func isNotWallOrFloor(paths map[Point]rune, fy int, p Point) bool {
	if p.y == fy {
		return false
	}

	return isNotWall(paths, p)
}

func simulateSandWithFloor(paths map[Point]rune, fy int) {
	p := Point{x: 500, y: 0}
	top := Point{x: 500, y: 0}
	ymin := 0
	for {
		if paths[top] == 'o' {
			return
		}

		np := p.down()
		if isNotWallOrFloor(paths, fy, np) {
			p = np
		} else {
			if isNotWallOrFloor(paths, fy, np.left()) {
				p = np.left()
			} else if isNotWallOrFloor(paths, fy, np.right()) {
				p = np.right()
			} else {
				paths[p] = 'o'
				if ymin > p.y {
					ymin = p.y
				}
				p = Point{x: 500, y: ymin - 1}
			}
		}
	}
}

func part2(paths map[Point]rune) int {
	ymax := 0
	for p := range paths {
		if ymax < p.y {
			ymax = p.y
		}
	}
	simulateSandWithFloor(paths, ymax+2)
	return countResting(paths)
}

func main() {
	lines := readInput()
	fmt.Println(part1(parsePaths(lines)))
	fmt.Println(part2(parsePaths(lines)))
}
